### ISSUE TITLE
Autorun only changed suites/tests 

### ISSUE THREAD

**Issue by matepek** [hbenl/vscode-test-explorer#52]
Hello,

I have a feature idea:

It would be nice to rerun only the 'changed' executable's suite ONLY.
It is not possible now.

So for example a possible solution could be:
TestAdapter interface having a
> readonly autorunChanged?: vscode.Event\<ids:string[]\>;

And I could fire it with the relevant suite/test id(s).
In the menu '...' we need another option like: **Run changed tests automatically**.

Note that I always have to send a `TestLoadFinishedEvent` event because I have to reload the changed executable. I think I should fire `autorunChanged` after it. (?)

Also the `TestLoadFinishedEvent` will triggers your autorun mechanism, but that feature and the new one could live next to each other. It could cause that tests are run twice if both feature are enabled, but it can be the user's choice. 

Or... `TestLoadFinishedEvent` interface could have a 
> changedIds: string[] //or similar

field. In this case you could see on time that you need to autorun by yourself or not.

I would prefer the first option, because I would like to add a `dependsOn:string[]` property. And in case of a change I would trigger the change of the related id(s).

Bests,

**Comment by matepek** [comment]
Also a configuration for this to be auto-enabled would be neat. (like [this](https://github.com/hbenl/vscode-test-explorer/issues/51))

**Comment by hbenl** [comment]
I've also been thinking about this (I want to add some code analysis to my adapters to detect the dependencies in the project - it can't catch all dependencies (due to dependency injection or javascript's dynamic `require`), but it should work in many if not most projects without the user having to manually specify the dependencies like with your `dependsOn` property).

Here's what I currently think about doing:
* tie the autorun feature to the retire feature: when the user enables autorun for some tests, then these tests will be run whenever they are retired (manually or through some adapter event). I think this really makes sense given the intended purpose of these features.
* replace the `autorun: vscode.Event<void>` event with a new `retire: vscode.Event<string[] | undefined>` event (I figured that autorun is a bad name as this event just informs the explorer that some test states are outdated and it's up to the user to decide if he wants to autorun in this case) just like your proposed `autorunChanged` event (except that you can call it without an argument to retire all tests).
* adapters should send a `retire` event immediately after the `TestLoadFinishedEvent` to inform the explorer which tests have changed. Furthermore, they should use the `retire` event if the sources of the app have changed and if an adapter knows which tests may be affected by the change, it can retire just those tests (otherwise it would send a `retire` event without an argument, which retires all tests).
* if the user enables autorun for all tests then the explorer will run only the retired tests, trusting the adapter that the other tests don't need to be run.
* if an adapter doesn't send `retire` events then the user can still configure the explorer to retire all tests after a `TestLoadFinishedEvent`.

**Comment by matepek** [comment]
Sounds logical.

There is a situation, which have to be handled somehow I guess.
Let's say we have 2 suites. _Suite1_ depends on _1.txt_ and _1-2.txt_. _Suite2_ depends on _1-2.txt_ and _2.txt_.
They will be watched with a file system watcher. A rebuild touches all the 3 txt files.
So now if I rebuild I will get three different events in a short interval: Let's say the order is 1, 1-2, 2.
If I only just forward the events, the _Test Explorer_ will get the following retire events in this order:

1. [\<suite1 id\>]
2. [\<suite1 id\>, \<suite2 id\>]
3. [\<suite2 id\>]

I think one of us needs to implement a logic which accumulates the ids. If you do it then other users won't have to. If we do it, your logic is more general.

Let's see a sketch what I am thinking about: (I assume we have a `retireDelayInSec` config.)
```ts
const delayBeforeAutorunRetired = 2; // it should be a config

const data = new class {
  public idBuffer: Set<string> | undefined = new Set();
  public lastArrivedAt: number | undefined = undefined; // undefined means: not scheduled
}();

const s = (): void => {
  if (data.lastArrivedAt === undefined) throw new Error('unexpeced'); // shouldn't be here

  if (Date.now() - data.lastArrivedAt >= retireDelayInSec * 1000) {
    const ids: string[] | undefined = data.idBuffer === undefined ? undefined : [...data.idBuffer.values()];

    data.idBuffer = new Set();
    data.lastArrivedAt = undefined;

    // dosomething(ids)
  } else {
    setTimeout(s, delayBeforeAutorunRetired * 1000 - (Date.now() - data.lastArrivedAt));
  }
};

adapter.retire((tests:string[]|undefined)=>{
  if (tests === undefined) {
    data.idBuffer = undefined;
  } else if (data.idBuffer !== undefined) {
    for (const test of tests) {
      data.idBuffer.add(test);
    }
  }

  if (data.lastArrivedAt === undefined) {
    setTimeout(s, delayBeforeAutorunRetired);
  }

  data.lastArrivedAt = Date.now();
});
```
You got the idea...

In case of my sketch there is a situation when we lose information: 

1. retire.fire([\<an id of a skipped test\>])
2. retire.fire(undefined)

This is interesting for me because my extension runs skipped tests if their ids are specifically mentioned in the `run([...])` array.
But in the example the second event will override the idBuffer and because the first event's id is a skipped one it won't be run.
But I don't mind it. It is a sharp corner case I guess.

Will you accept events while test are running or reloading? Or shall we wait until the "End" events before sending the retire ones?

I'm looking forward for the feature.

**Comment by hbenl** [comment]
I've just published version 2.9.0, which adds support for the `retired` event as described above (except for one little detail: the compatibility behavior that retires and autoruns all tests after a `TestLoadFinishedEvent` is automatically enabled if the adapter doesn't implement the retire event).

> There is a situation, which have to be handled somehow I guess [...]

I wanted to add this functionality to the `vscode-test-adapter-util` package (hence the Test Explorer 
logic remains general *and* adapter implementors don't need to implement it themselves) but then I 
realized that such a mechanism is already out there: just use the [debounce-collect](https://www.npmjs.com/package/debounce-collect) package to debounce and collect the `RetiredEvent`s, all that remains for you to do is to merge the collected `RetiredEvent`s into one and send that to the Test Explorer.

> Will you accept events while test are running or reloading? Or shall we wait until the "End" events before sending the retire ones?

Currently there is no logic in the Test Explorer to deal with this.
Eventually I want to add some logic that schedules test loads and runs, i.e. the events are always accepted but if an event is received while running or reloading, the event will be deferred until after that process has finished.

**Comment by matepek** [comment]
I guess we don’t have to reinvent the wheel. :)
